universal_transfer_operator/src/universal_transfer_operator/data_providers/filesystem/google/cloud/gcs.py

```diff
@@ -0,0 +1,160 @@
+        transfer_mode: TransferMode = TransferMode.NONNATIVE,
+    ):
+        super().__init__(
+            dataset=dataset,
+            transfer_params=transfer_params,
+            transfer_mode=transfer_mode,
+        )
+        self.transfer_mapping = {
+            Location.S3,
+            Location.GS,
+        }
+
+    @cached_property
+    def hook(self) -> GCSHook:
+        """Return an instance of the database-specific Airflow hook."""
+        return GCSHook(
+            gcp_conn_id=self.dataset.conn_id,
+            delegate_to=self.delegate_to,
+            impersonation_chain=self.google_impersonation_chain,
+        )
+
+    def check_if_exists(self) -> bool:
+        """Return true if the dataset exists"""
+        return self.hook.exists(bucket_name=self.bucket_name, object_name=self.blob_name)
+
+    @contextmanager
+    def read(self) -> list[TempFile]:
+        """Read the file from dataset and write to local file location"""
+        if not self.check_if_exists():
+            raise ValueError(f"{self.dataset.path} doesn't exits")
+
+        logging.info(
+            "Getting list of the files. Bucket: %s; Delimiter: %s; Prefix: %s",
+            self.bucket_name,  # type: ignore
+            self.delimiter,
+            self.prefix,
+        )
+        files = self.hook.list(
+            bucket_name=self.bucket_name,  # type: ignore
+            prefix=self.prefix,
+            delimiter=self.delimiter,
+        )
+
+        try:
+            local_file_paths = []
+            if files:
+                for file in files:
+                    _, _, file_name = file.rpartition("/")
+                    with NamedTemporaryFile(suffix=file_name, delete=False) as tmp_file:
+                        self.hook.download(
+                            bucket_name=self.bucket_name,
+                            object_name=self.blob_name,
+                            filename=tmp_file.name,
+                        )
+                        local_file_paths.append(TempFile(tmp_file=tmp_file, actual_filename=file_name))
+
+            yield local_file_paths
+        finally:
+            # Clean up the local files
+            self.cleanup(local_file_paths)
+
+    def write(self, source_ref: list[TempFile]) -> list[str]:
+        """Write the file from local file location to the dataset"""
+        bucket_name = self.bucket_name
+        object_prefix = self.blob_name
+
+        destination_objects = []
+        if source_ref:
+            for file in source_ref:
+                # There will always be a '/' before file because it is
+                # enforced at instantiation time
+                dest_gcs_object = object_prefix + file
+                self.hook.upload(
+                    bucket_name=bucket_name,
+                    object_name=dest_gcs_object,
+                    filename=file.name,
+                    gzip=self.gzip,
+                )
+                destination_objects.append(dest_gcs_object)
```

Comment:
same here

Reply:
https://github.com/astronomer/astro-sdk/pull/1492/commits/29e5ffaf685e27c9b5da3db1659b3442f78f68c9 and https://github.com/astronomer/astro-sdk/pull/1492/commits/535465e155915816db932fb678a30687a400e304